lib/dry/types/map.rb

```diff
@@ -0,0 +1,110 @@
+module Dry
+  module Types
+    class Map < Definition
+
+      def initialize(keys: Undefined, values: Undefined, meta: EMPTY_HASH)
+        super(::Hash, keys: keys, values: values, meta: meta)
+        validate_options!
+      end
+
+      # @return [Type, Undefined]
+      def keys
+        options[:keys]
+      end
+
+      # @return [Type, Undefined]
+      def values
+        options[:values]
+      end
+
+      # @return [String]
+      def name
+        "Map"
+      end
+
+      # @param [Hash] hash
+      # @return [Hash]
+      def call(hash)
+        try(hash) do |failure|
```

Comment:
Following how `Array` and `Hash` are designed:
- the method `call` simple call the `coerce` which return the output or an Error
- the method `try` would also the `coerce` method and depending on the result it should return `Result {Success| Failure }` object and only yield the `Result::Failure` if a block was given

Maybe we can stick to that convention? WDYT? @fledman

Reply:
So in my implementation, `coerce` is more similar to `Schema#resolve`, in that it always returns a `Result` object.

Then `call`/`try`/`valid?` build off of that to behave the same way as the convention:
  * call: return the output or raise an error
  * try:  yield or return the `Result` object
  * valid?: result.success?

I organized the code this way to cleanly separate the actual logic from error handling + output formatting.

Reply:
More specifically, `Schema` implements `try` using `call`, while `Map` implements `call` using `try`. 

This reversal allows us avoid using raise/rescue for internal error handling e.g. https://github.com/dry-rb/dry-types/blob/f3c2561dbad3b1d44b564951c823983fdf30493d/lib/dry/types/hash/schema.rb#L56-L66

Reply:
@fledman Thanks for the clarification. I'm not having a strong opinion about this, what I'm saying I'm ok with both solutions. I have created a gist https://gist.github.com/GustavoCaso/5376d5a0b5c195ec05c29464868543c2 just you can see the result, also I think the method `try` with this solution is more readable 😄.

Once again thank you for working on this ❤️

Reply:
I generally prefer pushing raised errors to the edge of library code, and using other mechanisms  for internal handling.

But I'm not the owner here, so it's up to you.